src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt

```diff
@@ -75,80 +79,98 @@ class NexusPublishPlugin : Plugin<Project> {
         rootProject.tasks.withType(AbstractTransitionNexusStagingRepositoryTask::class.java).configureEach {
-            transitionCheckOptions.convention(extension.transitionCheckOptions)
+            this.transitionCheckOptions.convention(extension.transitionCheckOptions)
+            this.stagingRepositoryId.convention(registry.map { it[repository.get().name].stagingRepositoryId })
         }
         extension.repositories.all {
-            username.convention(rootProject.provider { rootProject.findProperty("${name}Username") as? String })
-            password.convention(rootProject.provider { rootProject.findProperty("${name}Password") as? String })
-            publicationType.convention(PublicationType.MAVEN)
-
             val repository = this
-            val retrieveStagingProfileTask = rootProject.tasks.register<RetrieveStagingProfile>("retrieve${capitalizedName}StagingProfile", rootProject.objects, extension, repository)
+
+            repository.username.convention(rootProject.provider { rootProject.findProperty("${name}Username") as? String })
+            repository.password.convention(rootProject.provider { rootProject.findProperty("${name}Password") as? String })
+            repository.publicationType.convention(PublicationType.MAVEN)
+
+            @Suppress("UNUSED_VARIABLE") // Keep it consistent.
+            val retrieveStagingProfileTask = rootProject.tasks.register<RetrieveStagingProfile>(
+                "retrieve${capitalizedName}StagingProfile"
+            ) {
+                this.group = PublishingPlugin.PUBLISH_TASK_GROUP
+                this.description = "Gets and displays a staging profile id for a given repository and package group. " +
+                    "This is a diagnostic task to get the value and " +
+                    "put it into the NexusRepository configuration closure as stagingProfileId."
+                this.repository.convention(repository)
+                this.packageGroup.convention(extension.packageGroup)
+            }
             val initializeTask = rootProject.tasks.register<InitializeNexusStagingRepository>(
                 "initialize${capitalizedName}StagingRepository",
-                rootProject.objects,
-                extension,
-                repository,
                 registry
             )
+            initializeTask {
+                this.group = PublishingPlugin.PUBLISH_TASK_GROUP
+                this.description = "Initializes the staging repository in '${repository.name}' Nexus instance."
+                this.repository.convention(repository)
+                this.packageGroup.convention(extension.packageGroup)
+            }
             val findStagingRepository = rootProject.tasks.register<FindStagingRepository>(
                 "find${capitalizedName}StagingRepository",
-                rootProject.objects,
-                extension,
-                repository,
                 registry
             )
             findStagingRepository {
```

Comment:
My goal was to merge each of these into:
```
val name = rootProject.tasks.register<T>("name") { ... }
```
but these last two I can't figure out how to remove the registry.

I think the answer is using the build service (injecting it inside the task), but that's only possible from Gradle 6.1 (see `NexusPublishPlugin.createRegistry`) and the current minimum is 6.0.

@3flex any ideas?

Reply:
Just moved from unannotated constructor property to @Internal managed property: https://github.com/gradle-nexus/publish-plugin/pull/242/commits/e964697bcaf1e996feea3fc319df06777445bba2